File: android/src/main/java/com/circomwitnesscalc/CircomWitnesscalcModule.java
package com.circomwitnesscalc;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.module.annotations.ReactModule;

import android.util.Base64;

import io.iden3.circomwitnesscalc.WitnessCalcKt;
import io.iden3.circomwitnesscalc.WitnesscalcError;


@ReactModule(name = CircomWitnesscalcModule.NAME)
public class CircomWitnesscalcModule extends ReactContextBaseJavaModule {
  public static final String NAME = "CircomWitnesscalc";

  public CircomWitnesscalcModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  @ReactMethod
  public void calculateWitness(String inputs, String graph, Promise promise) {
    try {
      // Decode base64
      byte[] graphBytes = Base64.decode(graph, Base64.DEFAULT);

      byte[] witness = WitnessCalcKt.calculateWitness(
        inputs,
        graphBytes,
        128
      );

      // Encode base64
      String witnessBase64 = Base64.encodeToString(witness, Base64.DEFAULT);

      promise.resolve(witnessBase64);
    } catch (WitnesscalcError e) {
      promise.reject(String.valueOf(e.getCode()), e.getMessage());
    }
  }
}

File: android/src/main/java/com/circomwitnesscalc/CircomWitnesscalcModule.kt
package com.circomwitnesscalc

import android.util.Base64
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import io.iden3.circomwitnesscalc.WitnesscalcError
import io.iden3.circomwitnesscalc.calculateWitness

class CircomWitnesscalcModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String {
    return NAME
  }

  // Example method
  // See https://reactnative.dev/docs/native-modules-android
  @ReactMethod
  fun calculateWitness(inputs: String, graph: String, promise: Promise) {
    try {
      // Decode base64
      val graphBytes = Base64.decode(graph, Base64.DEFAULT);

      val witness = calculateWitness(
        inputs,
        graphBytes
      )

      // Encode base64
      val witnessBase64 = Base64.encodeToString(witness, Base64.DEFAULT);

      promise.resolve(witnessBase64);
    } catch (e: WitnesscalcError) {
      promise.reject(e.code.toString(), e.message);
    }
  }

  companion object {
    const val NAME = "CircomWitnesscalc"
  }
}
